### ISSUE TITLE
compiler: Document side effects when adding documents vs adding executables

### ISSUE THREAD

**Issue by erikwrede** [apollographql/apollo-rs#442]
When using the compiler with a single GQL-Schema added via `add_type_system` and multiple GQL-Operations that are added after an initial validation of the schema, there can be significant performance differences in the validation of the operations  depending on the procedure. 

According to the documentation, `add_document` should behave similarly to `add_executable`, but with the added ability to accept SDL input. However, it seems to trigger re-validation of the initially validated schema, even without any new types added, which can be costly for large schemas.

For my testing & benchmarking, I used the [GitLab schema](https://gitlab.com/api/graphql) with 60000 lines of SDL. Validation using `apollo-compiler` takes `~ 40 ms` on my machine. Validating a 250 line nested query with inputs and variables against this schema takes `~1ms`. If I use `add_document` for a query against this schema, the operation validation time increases to `~40ms`. I have attached a simple example below. 

The issue seems to be the definition of  `type_definition_files` in the `InputDatabase`.  Since at this stage, all of the documents should be unparsed sources, there is no way to exclude them from the type system in case the root document node doesn't contain any types, am I reading that correctly?

If this is the intended behavior, it would be very helpful to clarify the possible side effects in the documentation or adjust the nomenclature. 
Intuitively, every executable is a document node with definition nodes consisting of only operation and fragment nodes, so there should be no issues with `add_document` in this case.

---
<details>
  <summary>Repro</summary>

```rust
fn bench_simple_query_compiler_with_validation() {
    let query = include_str!("testdata/simple_query.graphql");
    let schema = include_str!("testdata/simple_schema.graphql");

    let mut compiler = ApolloCompiler::new();
    compiler.add_type_system(schema, "schema.graphql");

    compiler.db.object_types();
    compiler.validate();

    let start = Instant::now();

    let file_id = compiler.add_document(query, ""); // FIXME use add_executable instead

    let elapsed_parsing = start.elapsed();
    let mut diagnostics = compiler.db.ast(file_id)
        .errors()
        .into_iter()
        .map(|err| {
            ApolloDiagnostic::SyntaxError(SyntaxError {
                src: compiler.db.source_code(file_id),
                span: (err.index(), err.data().len()).into(),
                message: err.message().into(),
            })
        }).collect::<Vec<ApolloDiagnostic>>();

    let elapsed_parser_syntax_check = start.elapsed();
    // We only want to validate the new operation against the schema here, everything else is already validated
    diagnostics.extend(compiler.db.validate_executable(file_id));

    let elapsed = start.elapsed();

    println!("Time elapsed after parsing is: {:?}", elapsed_parsing);
    println!("Time elapsed after syntax error check is: {:?}", elapsed_parser_syntax_check);
    println!("Time elapsed after parsing & validation is: {:?}", elapsed);
}
```
</details>

**Comment by lrlna** [comment]
oooofff, that sounds like a bug actually! Type system re-validation should not be triggered for an already validated type system, especially since the original type system was not updated. Under the hood `salsa` caches all internal queries given their input stays the same, and since the inputs for the type system validation queries has not changed, we should be incurring the cost of additional parsing. 

Going to investigate this a bit more!

**Comment by erikwrede** [comment]
The input for the type system queries has indeed changed since they're all relying on aggregating typed from `type_definition_files`, as mentioned above. I'm not sure if salsa can cache that kind of case, since all underlying inputs change  but the change doesn't impact the output of the computation (because no types were added)

**Comment by SimonSapin** [comment]
I expect this should be fixed in apollo-compiler 1.0: the public API no longer has compiler objects. Schema validation and executable validation should now be completely separate.

If you have both a schema and a query in the same file, `add_document` is replaced by [`parse_mixed`](https://docs.rs/apollo-compiler/1.0.0-beta.1/apollo_compiler/fn.parse_mixed.html) which returns a tuple of `Schema` and `ExecutableDocument` whose respective validation is still separate.

PR https://github.com/apollographql/apollo-rs/pull/676 adds a rough port of your repro code. I didn’t have your 250 lines query so I ran it with a very simple one instead. Please let us know if you still find something suspicious!

**Comment by lrlna** [comment]
I will close this issue, as we've redesigned the compiler with the `1.0.0-beta.x` releases and this side effect is no longer there. Feel free to reopen if something isn't right still.
